### ISSUE TITLE
Find or create a database of all politicians

### ISSUE THREAD

**Issue by Irio** [datasciencebr/serenata-de-amor#196]
As mentioned [in articles](https://datasciencebr.com/expandindo-a-serenata-de-amor-para-outras-esferas-48bed204d36d), although Serenata de Amor is currently focused in the Chamber of Deputies, the same models could be adapted for other Brazilian quotas, like Senate, Legislative Chambers, City Halls... Having a database of whole Politicians life - which is public information - can be very useful for exploration and Supervised Machine Learning. Related hypotheses are many: is a suspicious expense made in a place where the politician has a lot of influence? Is the politician in the first term or we could infer a significant influence in the Congress?

Example:

| Name | Full Name | Position | Place | Inauguration | Ending |
|------|-----------|----------|-------|-------|-----|
| Fernando Collor | Fernando Affonso Collor de Mello | Mayor | Maceió - AL | 1979-01-01 | 1983-01-01 |
| Fernando Collor | Fernando Affonso Collor de Mello | Federal Deputy | AL | 1983-02-01 | 1987-02-01 |
| Fernando Collor | Fernando Affonso Collor de Mello | State Governor | AL | 1987-03-15 | 1989-05-14 |
| Fernando Collor | Fernando Affonso Collor de Mello | President | Brazil | 1990-03-15 | 1992-12-29 |
| Fernando Collor | Fernando Affonso Collor de Mello | Senator | AL | 2007-02-01 |  |

May be appealing to scrape Wikipedia for that, but I wish we could use [TSE](http://www.tse.jus.br/) or other governmental sources. Candidacies is also a very important dataset to be collected next.

**Comment by vitorbaptista** [comment]
http://everypolitician.org/ aims to be a database for all politicians across the globe, and could be an interesting source/partner. For instance, this is the data they have for Brazilian politicians: http://everypolitician.org/brazil/.

**Comment by allantorres** [comment]
Great data but they don't have the dates but is a start, 
TSE have all data we need : http://www.tse.jus.br/eleicoes/estatisticas/repositorio-de-dados-eleitorais
But we are going to need to clean up the data ...

**Comment by ppKrauss** [comment]
Hi, I am new here... About the suggestions, and the ID problem.  I prefer [*transparent* identifiers](https://en.wikipedia.org/wiki/Lex_(URN)#Transparent_identifiers) (as an URN of `completeName-birthDate-birthCity` or its [hash](https://en.wikipedia.org/wiki/Cryptographic_hash_function)), based on essential metadata. *Opaque* IDs like  [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier) are error-prone (essential metadata duplications)... Of course, when we have access to an official opaque ID like [CPF](https://en.wikipedia.org/wiki/Cadastro_de_Pessoas_F%C3%ADsicas), that is unique (RG is not unique), it is also perfect. So,

* [everypolitician.org](http://everypolitician.org/brazil/) have a big problem, the randomic ID.
* [TSE](http://www.tse.jus.br/eleicoes/estatisticas/repositorio-de-dados-eleitorais) have all data (!), as I checked, CPF is there. 

To clean up the TSE data: see LEIAME.pdf, some years are different... after `unzip` it, example of 2016-SP
* `iconv -f ISO-8859-1  -t UTF-8 consulta_cand_2016_SP.txt`
* it is a [CSV](https://en.wikipedia.org/wiki/Comma-separated_values) file with separator `;` without  header line. 
* Starting at field1, the *name* is `field11` and  CPF is the `field14`.
* All candidates of all years ... suppose an SQL database with all CSV imported into `all_tse` table, will be `SELECT DISTINCT field11 as name, field14 as cpf FROM all_tse`

**Comment by rafonseca** [comment]
Hi guys, I am also new here. I can contribute to this issue using jupyter, pandas and TSE data. I can start with the candicacies file. From that we can compose easily with pandas a table like Irio requested. We can further include campaing contributors data from TSE! Should I upload a notebook, a dataset or both?

**Comment by cuducos** [comment]
Sounds great, @rafonseca! Have you had the chance to get started? I'm interested in this topic, let me known if you're interested in pairing or something ; )

**Comment by silviodc** [comment]
Hi guys, after a brief look on chamber of deputes website I found the data about all deputes as hereafter:
```
numLegislatura>55
email>dep.fabiofaria@camara.gov.br
nomeProfissao>Administrador de Empresas
dataNascimento>01/09/1977
dataFalecimento>
ufRepresentacaoAtual>RN</ufRepresentacaoAtual>
situacaoNaLegislaturaAtual>Em Exercício
ideCadastro>141428
idParlamentarDeprecated>4460
nomeParlamentarAtual>FÁBIO FARIA
nomeCivil>FÁBIO SALUSTINO MESQUITA DE FARIA
sexo>M
partidoAtual>PSD
nome>Partido Social Democrático
gabinete>706
anexo>4
telefone>3215-5706
```
It is easy to catch, we only have to use their end-point, here:
http://www2.camara.leg.br/transparencia/dados-abertos/dados-abertos-legislativo/webservices/deputados/obterdeputados

Url to the mentioned example:
[FÁBIO FARIA](http://www.camara.leg.br/SitCamaraWS/Deputados.asmx/ObterDetalhesDeputado?ideCadastro=141428&numLegislatura=)

It is also possible to take detailed information about this deputes using the following url:
http://www.camara.leg.br/SitCamaraWS/Deputados.asmx/ObterDetalhesDeputado
The returned data is:
```
email 	String 	Email do parlamentar
nomeProfissao 	String 	Profissão do parlamentar
dataNascimento 	String 	Data de nascimento
dataFalecimento 	String 	Data de falecimento
ufRepresentacaoAtual 	String 	UF da representação parlamentar
situacaoNaLegislaturaAtual 	string 	Situação do parlamentar na legislatura (Em exercício, Licenciado, Afastado, Vacância e Suplência)
ideCadastro 		Identificado único do parlamentar
nomeParlamentarAtual 		Nome parlamentar
nomeCivil 		Nome Civil
sexo 		Sexo
partidoAtual 		LIST
gabinete 		LIST
comissoes 		LIST<comissao>
cargosComissoes 		LIST<cargoComissoes>
periodosExercicio 		LIST<periodoExercicio>
historicoNomeParlamentar 		LIST
filiacoesPartidarias 		LIST<filiacaoPartidaria>
historicoLider 		LIST<itemHistoricoLider>
```
Could we start from here?

**Comment by cuducos** [comment]
Hi @silviodc — many thanks for your interest.

However your proposal is only focused on the Chamber of Deputies (not _all politicians_ as asked by @irio) and, even more restrict, only covering the deputies in charge (from the link you sent: _Retorna os deputados __em exercício__ na Câmara dos Deputados_).

What we are looking for in this issue is data about all politicians: posts they had before getting into the Chamber of Deputies as well as where they headed afterwards, for example.

Maybe with this API you can contribute/enhance other parts os the project, such as the scripts that scrap politicians full name, for example.

**Comment by rafonseca** [comment]
Hi @cuducos.
Yes, I have started it. I have a Jupyter notebook that download candidacies data from TSE and import to pandas setting the correct header. They have three different versions for the headers. For the moment, we have a list of all candidates from 2004 to 2016 with useful information like if the politician was elected or not. 
TSE informs that they are replacing the files from 1994 to 2002. Indeed, some of these files do not follow the format described in their pdf. For the year before 1994, they have no information about the politicians.
Yes, we could do something together. What is the proper way to share my code with you? That may sound strange, but I have been working alone for years and all this github stuff is new for me. I am sure you can help me :)

**Comment by ppKrauss** [comment]
Hi @rafonseca, good, let's use TSE database (!). 
see in the TSE zip file  the `LEIAME.pdf`, with instructions about  each range of years... 
See [my post above](https://github.com/datasciencebr/serenata-de-amor/issues/196#issuecomment-283912395), no need of statistical analysis (Pandas), the data are reliable. What we need is to join all CSVs data, and filter it with standard SQL DISTINCT.

About Github stuff, you have a login, it is a first step (!), goo to https://github.com/rafonseca?tab=repositories and click in the green big button  "NEW".

**Comment by rafonseca** [comment]
Hi @ppKrauss  and @cuducos.
For the moment, I created a new repository in my account called tse-data and I put the notebook there. Maybe I should have created a branch on my fork of serenata ? Anyway, for now you can see what I did on the notebook.
@ppKrauss, I have used pandas just to import/export csv, join and filter the tables. Sorry, I don't know SQL. I am having problem with the years 2004 and 2008. Need to check it. Maybe SQL can do a better job there?
At the end of the notebook, we have a table similar to that requested by @Irio , but the dates are not so precise.

**Comment by jtemporal** [comment]
Hi @rafonseca it is very good that you want to contribute!
>What is the proper way to share my code with you?

Though a branch on the fork of the project would sound ideal, you can always [link the code](https://github.com/rafonseca/tse-data/blob/master/collect_and_make_csv.ipynb) you have on your repo here in the comment and move that code to a branch later (specially if you want to make a PR \o/)

> What we need is to join all CSVs data, and filter it with standard SQL DISTINCT.

You don't exactly need SQL to do what @ppKrauss suggested, pandas should be enough to achieve that goal, I know I had to do similar csv operations and it worked just fine for me ;)

> I am having problem with the years 2004 and 2008

maybe you could explain a little bit more about the problem that you are having?

**Comment by rafonseca** [comment]
Hi @jtemporal 
wow, we can link the code here and it displays the notebook. Wonderful! Thanks. I really needed this information.
About the years 2004 and 2008, I have just checked. It is not a big issue. Just need to convert some '#NI#' to NaN if I want those columns to be int.

**Comment by cuducos** [comment]
> You don't necessarily need SQL to do what @ppKrauss suggested, pandas should be enough to achieve that goal, I know I had to do similar csv operations and it worked just fine for me ;

I think @ppKrauss was merely using `DISTINCT` to express his idea, not to point us towards `SQL`. Anyway I'm still not sure about the benefits of something in the lines of `SELECT DISTINCT field11 as name, field14 as cpf` — for me the point is to have information about the political career of each politician, so it's totally fine to have more than one line per `name` and `cpf` (just check @Irio's [example](https://github.com/datasciencebr/serenata-de-amor/issues/196#issue-211070445), all these lines are from the same politician). Can you clarify that, @ppKrauss? It's not useful to have all candidacies of all years, but the info about when each politician was in the office, and where (which office).

@rafonseca — I think @jtemporal already gave you some tips. Now you can link to your notebook/repo and raise specific doubts ; ) We're here to help you!

Many thanks you all

**Comment by ppKrauss** [comment]
Hi all, thanks @cuducos, @jtemporal  and @rafonseca. 

**Summary**: let's get all `<name,cpf>` first, I can help @rafonseca  at [his `tse-data` project](https://github.com/rafonseca/tse-data/issues). 

 -----

Answering in detail:

* yes @jtemporal  and  @cuducos the SQL here is only a  jargon to express the ideia.

* about clarify request of @cuducos, *"It's not useful to have all candidacies of all years, but the info about when each politician was in the office, and where"*. My view is that the best is normalized data, 

   *  **name** (full name) and **CPF** fields (simple `<name,cpf>` table), are central to define unically the [public identity of Brazilian people](https://medium.com/@peterkrauss/sobre-identificadores-ids-e-nomes-e-o-direito-de-saber-o-nome-do-outro-a264ff1fd7c#.weaavkyqq). This table can be maintained by a little  "consortiun of projects", and I can collaborate as [OKBR member](https://br.okfn.org/membros/). <br/>PS: the CPF in this case is a public ID, no problem about "transparent ID". This table also need, perhaps, "source of information" and "last updated in the source" fields. 

   * other fields can be maintained into a JSON field (I am using  [PostgreSQL9.5+ with  JSONb](https://www.postgresql.org/docs/current/static/functions-json.html))... Or separate CSV file (or SQL table) normalized by CPF9 (CPF can be casted to a *long integer* ID using first 9-digits), to not repeat *full name* and associated fields. <br>Imagining `<cpf, position, place, inauguration, ending>` as @Irio's illustration.

* (**next discusstion** after get data) ideal database or format?  Example: [LexML](http://lexml.gov.br/)  (an [ePING](http://eping.governoeletronico.gov.br/) standard) use [URN LEX](https://en.wikipedia.org/wiki/Lex_(URN))... The "jurisdiction fragment" is a goot *transparent ID* for place field. Instead to use IBGE opaque IDs, we can adopt LexML transparent ones. <br/>"Federal"=`br`, "Estadual, SP"=`br;sp` or `br.sao.paulo`, and  "municipal, de Campinas"=`br;sp;campinas`. 

* **(for future)** the @vitorbaptista suggestion is good... Perhaps we can collaborate with [everypolitician.org/brazil](http://everypolitician.org/brazil/) database, exporting our data to there.  [Here](http://docs.everypolitician.org/data_structure.html) and at  [Popolo's classes and properties](http://www.popoloproject.com/specs/person.html#classes-and-properties) a good structure specification... My only objection, as comment [above](#issuecomment-283912395), is about UUID use,  is better to use CPF=[vatID](http://schema.org/vatID) (or a *transparent ID* convention)... As collective we can discuss about IDs there.

... Seems complex? let's start with hands-on at @rafonseca's repo... And back to the discusion here with something simple and concrete.

**Comment by rafonseca** [comment]
Hey guys, the notebook is updated [here](https://github.com/rafonseca/tse-data/blob/master/collect_and_make_csv.ipynb) Wow!
The first time you run it, you'll want to run all cells. The next time you skip the download and the unzip part. From all the columns provided, I selected (in sel_columns) 12 to stay with a reasonable csv file size. But it is easy to select more columns. Just uncomment there.
On python 2.7, I cannot export to csv using compression and encoding together. So, actually, I didnt try to export.
At section 'only elected politicians', I filtered the dataframe to obtain only elected politicians and the essential columns. Take a look in the last cell.

**Comment by ppKrauss** [comment]
Hi @rafonseca , my suggestion of first-step is to create all basic CSV (filtered) files, that is, all  `<name,cpf>` tables, one for each year... See [this example at datapackage-viewer](http://data.okfn.org/tools/view?url=https%3A%2F%2Fraw.githubusercontent.com%2Fdatasets-br%2Fpublic-person%2Fmaster%2Fdatapackage.json).

**Comment by rafonseca** [comment]
Hello @cuducos, @jtemporal, @ppKrauss and other passing by!

First of all, it seems it is missing a point in this discussion. @ppKrauss is pointing us to a robust database of politicians to be used outside Serenata while others like me are thinking in a solution to use inside Serenata. That's why there are divergences about how the database should be structured. I have limited time, and so, I will restrict my collaboration to Serenata project.

I updated the [notebook](https://github.com/rafonseca/tse-data/blob/master/collect_and_make_csv.ipynb) again. The purpose of this notebook is to generate TSE_politicians.csv and TSE_candidacies.csv. I restricted the columns to the following informations: CPF, name, function, city, state, year. TSE_politicians is a subset of TSE_candidacies based on the column DESC_SIT_TOT_TURNO. The filtering that defines the subset need some improvements. For example, I dont know if 'SUPLENTE' should always be included in TSE_politicians.
@Irio, do you think is useful to have both datasets, considering that one is a subset of the other?

@ppKrauss, at the end of this notebook I've included some lines that generate the csv files, one for each year, as you have requested.

In the near future, I intend to transform the notebook in a proper script following the coding guidelines, but I need to install python 3 before. I have some projects running on python 2 and I am afraid of causing conflicts...

**Comment by ppKrauss** [comment]
Hi all, this is a first draft, I will publish with all data at git in the next week.

## TSE-candidatos data analysis

There are 3.5 million (3520846) of CSV rows to be checked over uniqueness of person, it is expected that records are repeated for each election year. 

* all (100%) have [name](https://schema.org/name) (full name) valid property.
* 59% (2081277 rows) have [birthDate](https://schema.org/birthDate) valid property.
* 43% (1523871 rows) have [vatID](https://schema.org/vatID) ([Brazilian CPF](https://www.wikidata.org/wiki/Q5016244)) valid property.

Profile of the CSV files: 

* There are 310 source files `consulta_cand*.txt` (state and year), and the its fields are not uniform.

* Supposing uniformity, the relevant fields are at positions 3, 6, 11, 14 and 27 (from 1), and secondary positions 15, 28 and 31. Supposing corresponds respectivally to fields (of the LEIAME.PDF documentation) `ANO_ELEICAO`,`SIGLA_UF`,`NOME_CANDIDATO`,`CPF_CANDIDATO`,`DATA_NASCIMENTO`,  `NOME_URNA_CANDIDATO`,`NUM_TITULO_ELEITORAL_CANDIDATO`,`SEXO`.

* Supposing "newer is better", so, when repeat we can select only the last version.

I am using pure SQL... After import all CSV lines to one table, filtering the  rows with  valid name, valid birth and valid CPF-checksum, removing duplicates, remains **1538241** rows:

* 89% (1375305 rows) have distinct names;
* ... There are some little of names with problems, as  "0DÁRICO", "0SEAS", and "ABRRAÃO".
* 89% (1375973 rows) have distinct CPFs (remained 11% of 135485 rows are repeated CPF for distinct name-birth);
* ~1% (14270 rows) have empty CPF;
* ~100 rows  have two or more distinct CPFs to the same name and same birth date.
* ... Grouping by CPF: there are 21 cases of "CPF 99999999999" (invalid but a lie approved by TSE), some with more than 5 repetitions (eg. CPFs 02990529829  and 32844085253), 859 cases with more tham 3 repetitions, 11419 cases with more tham 2 repetitions.

Commom examples of same person with two names:

* *ortographic error*: Marli and Marly, both with birth date 1965-12-27 and CPF 45258007472. DELIZOM and DELIZON (CPF 21963541987), JOSE and JOSÉ (CPF 71404449949), etc. 

* *Real change of name* (eg. after marriage): "MARIA DO CARMO PRATES COSTA" and "MARIA DO CARMO PRATES MARQUES", both 1965-03-09 and CPF 37921800572. 

The homonyms:

*  Main examples: "MARIA APARECIDA DA SILVA" (259), "MARIA JOSE DA SILVA" and "MARIA APARECIDA DOS SANTOS". 

* Same name and same birthDate: very rare, but not impossible. MARIA APARECIDA DA SILVA  of 1975-08-01, CPFs 28189313800 and 84915099172.

* There are also cases of same  name, but less frequent name. Example to check perhaps is a father-and-son case, both "ABISAI DE OLIVEIRA SOUSA", "Vereador" at MA cities in 2016, [supposed junior](https://eleicoes2016.inf.br/candidato-vereador/abisai-sousa_100000005451) and  [supposed older akin](https://eleicoes2016.inf.br/candidato-vereador/abisai_100000012410).

**Comment by cuducos** [comment]
That's indeed a good overview of data.

Feel free to open a PR with a notebook exploring that and/or a script collecting this data — that would be awesome ; )

---

Just a friendly reminder: 

> I will publish with all data at git

Just a reminder: as stated in our [`CONTRIBUTING.md`](https://github.com/datasciencebr/serenata-de-amor/blob/master/CONTRIBUTING.md#best-practices) we do commit datasets to the git repository. What we do is add the script that gathers data to the repo (at `src/`), ok?

**Comment by ppKrauss** [comment]
Hi @cuducos , it is a first proposal of [OKBR-Serenata](https://br.okfn.org/2017/03/17/open-knowledge-brasil-apoia-a-operacao-serenata-de-amor/) work for a commom  back-end. The repo *is under construction*, [here](https://github.com/datasets-br/public-person).

**Comment by rafonseca** [comment]
Hi @cuducos,
I did a PR for the tse data collect script. Sorry for the delay. I have been busy last weeks.

**Comment by cuducos** [comment]
Closed by #208
